clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -225,11 +230,20 @@ public double compressionRate() {
     }
 
     public boolean isFull() {
-        return recordsBuilder.isFull();
+        return !isWritable || recordsBuilder.isFull();
```

Comment:
Hmm, in RecordAccumulator.append(), we return the following. So if the current batch doesn't have enough space, we will create another one. Then dq.size() will be > 1 and isFull will be true anyway.

`return new RecordAppendResult(future, dq.size() > 1 || batch.isFull(), true);`